sdk/provisioning/Azure.Provisioning/src/Infrastructure.cs

```diff
@@ -93,6 +94,97 @@ public virtual void Remove(Provisionable resource)
         }
     }
 
+    /// <summary>
+    /// Checks whether an name is a valid bicep identifier name comprised of
+    /// letters, digits, and underscores.
+    /// </summary>
+    /// <param name="identifierName">The proposed identifier name.</param>
+    /// <returns>Whether the name is a valid bicep identifier name.</returns>
+    public static bool IsValidIdentifierName(string? identifierName)
+    {
+        if (string.IsNullOrEmpty(identifierName)) { return false; }
+        if (char.IsDigit(identifierName![0])) { return false; }
+        foreach (char ch in identifierName)
+        {
+            if (!char.IsLetterOrDigit(ch) && ch != '_')
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /// <summary>
+    /// Validates whether a given bicep identifier name is correctly formed of
+    /// letters, numbers, and underscores.
+    /// </summary>
+    /// <param name="identifierName">The proposed bicep identifier name.</param>
+    /// <exception cref="ArgumentNullException">Throws if null.</exception>
+    /// <exception cref="ArgumentException">Throws if empty or invalid.</exception>
+    public static void ValidateIdentifierName(string? identifierName)
```

Comment:
Would it be possible to take a `string? parameterName = null` parameter? That way this method could be used by other callers and they can just let the ArgumentException through.

For example:

```C#
    public static IResourceBuilder<T> WithParameter<T>(this IResourceBuilder<T> builder, string name, IEnumerable<string> value)
        where T : AzureBicepResource
    {
        Infrastructure.ValidateIdentifierName(name, nameof(name));
        builder.Resource.Parameters[name] = value;

```

Reply:
I don't like adding things to the API that are primarily for my implementation benefit, but it looks like this is a pattern in a lot of the newer validation methods.  Changed.

Reply:
Note this isn't just for your implementation benefit. We are using this parameterName argument in dotnet/aspire:

https://github.com/dotnet/aspire/blob/a8b1a619d67549243342dfd2c92cfaf26905cb65/src/Aspire.Hosting.Azure/Utils/BicepIdentifierHelpers.cs#L13